### PR TITLE
Update Boost to 1.65 and make it default

### PR DESCRIPTION
**Pull request by abbradar**
###### Motivation for this change

Update the default package. It brings fixes for NVCC so that we can use new CUDA toolkit 9.0 with default Boost (which we need due to GCC 6 default). Also NumPy support is improved.

I've made packages that overrode Boost to a newer one use the default version. RStudio update is needed to fix build.

###### Things done

- [x] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `build-use-sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] Linux
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

I work on a system built with this Boost for some time, but don't have resources to test all packages.

### CODE REVIEW

File: pkgs/applications/editors/rstudio/default.nix [hunk 1]
```diff
@@ -1,20 +1,20 @@
...
-  version = "1.1.351";
+  version = "1.1.374";
   ginVer = "1.5";
   gwtVer = "2.7.0";
 in
 stdenv.mkDerivation rec {
   name = "RStudio-${version}";
 
-  buildInputs = [ cmake boost163 zlib openssl R qt5.full qt5.qtwebkit qt5.qtwebchannel libuuid unzip ant jdk makeWrapper pandoc ];
+  buildInputs = [ cmake boost zlib openssl R qt5.full qt5.qtwebkit qt5.qtwebchannel libuuid unzip ant jdk makeWrapper pandoc ];
```
Comment (peterhoeg) [inline pkgs/applications/editors/rstudio/default.nix]: You want to clean up the nativeBuildInputs while at it?

Comment (abbradar) [reply pkgs/applications/editors/rstudio/default.nix]: Good idea, done.

**Comment by abbradar** [comment]
Ugh, Github requested reviews automatically because I accidentially pushed the wrong branch. Sorry.

File: pkgs/applications/altcoins/default.nix [hunk 2]
```diff
@@ -1,4 +1,4 @@
-{ callPackage, boost155, boost162, boost163, openssl_1_1_0, haskellPackages, darwin, libsForQt5 }:
+{ callPackage, boost155, boost162, openssl_1_1_0, haskellPackages, darwin, libsForQt5 }:
```
Comment (peti) [inline pkgs/applications/altcoins/default.nix]: This build needs both boost155 and boost162?

**Review by peti — APPROVED** [review]
I like it. There's certainly going to be some fallout because of this change,so when the PR is merged there'll be another round of committing fixes to apps that break because of the new version. :-/

File: pkgs/applications/altcoins/default.nix (hunk 2 above)
Comment (abbradar) [reply pkgs/applications/altcoins/default.nix]: It's a subpackage tree, some altcoins seem to need older Boosts.

**Comment by abbradar** [comment]
Let's merge this in several days if noone objects; I'll be watching Hydra for possible fallout.

**Pull request merged**